apps/src/jsArray.js

```diff
@@ -0,0 +1,113 @@
+// Strict linting: Absorb into global config when possible
+/* jshint
+ unused: true,
+ eqeqeq: true,
+ maxlen: 120
+ */
+
+var codegen = require('./codegen');
+
+// Array.sort
+// iterative quicksort based on:
+// http://www.stoimen.com/blog/2010/06/18/friday-algorithms-iterative-quicksort/
+// (modified as a stateful native function to support callbacks)
+exports.generateSort = function (jsInterpreter) {
+  return function (opt_compFunc) {
+    if (this.length <= 1) {
+      return this;
+    }
+    var interpreter = jsInterpreter.interpreter;
+    var i;
+    var state = jsInterpreter.getCurrentState();
+    var useCompareFunc = (typeof opt_compFunc !== 'undefined') &&
+        (opt_compFunc !== interpreter.UNDEFINED);
+    if (typeof state.__i === 'undefined') {
+      state.__jsList = [];
+      for (i = 0; i < this.length; i++) {
+        state.__jsList[i] = this.properties[i];
+      }
+      if (useCompareFunc) {
+        state.__sortStack = [state.__jsList];
+        state.__sortedList = [];
+      }
+    }
+    if (!useCompareFunc) {
+      // No compare function supplied, use the native JS sort(), but supply
+      // a compare function that marshals the interpreter values back to native
+      // before comparing them.
+      state.__jsList.sort(function (a, b) {
+        var nativeA = codegen.marshalInterpreterToNative(interpreter, a);
+        var nativeB = codegen.marshalInterpreterToNative(interpreter, b);
+        return (nativeA < nativeB) ? -1 : (nativeA > nativeB ? 1 : 0);
```

Comment:
Every primitive in the interpreter is stored as an object. If we try to run the native `Array.sort` where all of the values are of type `Object`, I believe it will convert the objects to strings in order to compare, which will likely produce random results.